frappe/model/naming.py

```diff
@@ -38,25 +39,96 @@ def set_new_name(doc):
 		doc.run_method("autoname")
 
 	if not doc.name and autoname:
-		if autoname.startswith('field:'):
-			fieldname = autoname[6:]
-			doc.name = (doc.get(fieldname) or "").strip()
-			if not doc.name:
-				frappe.throw(_("{0} is required").format(doc.meta.get_label(fieldname)))
-				raise Exception('Name is required')
-		if autoname.startswith("naming_series:"):
-			set_name_by_naming_series(doc)
-		elif "#" in autoname:
-			doc.name = make_autoname(autoname, doc=doc)
-		elif autoname.lower()=='prompt':
-			# set from __newname in save.py
-			if not doc.name:
-				frappe.throw(_("Name not set via prompt"))
-
-	if not doc.name or autoname=='hash':
+		doc.name = _get_name_from_naming_options(autoname, doc)
+
+	# if the autoname option if 'field:' and no name was derived, we need to 
+	# notify
+	if autoname.startswith('field:') and not doc.name:
+		fieldname = autoname[6:]
+		frappe.throw(_("{0} is required").format(doc.meta.get_label(fieldname)))
+
+	# at this point, we fall back to name generation with the hash option
+	if not doc.name or autoname == 'hash':
 		doc.name = make_autoname('hash', doc.doctype)
 
-	doc.name = validate_name(doc.doctype, doc.name, frappe.get_meta(doc.doctype).get_field("name_case"))
+	doc.name = validate_name(
+					doc.doctype,
+					doc.name,
+					frappe.get_meta(doc.doctype).get_field("name_case")
+				)
+
+
+def _field_autoname(autoname, doc, skip_slicing=None):
+	"""
+	Generate a name using `DocType` field. This is called when the doctype's 
+	`autoname` field starts with 'field:'
+	"""
+	fieldname = autoname if skip_slicing else autoname[6:]
+	name = (doc.get(fieldname) or '').strip()
+	return name
+
+
+def _prompt_autoname(autoname, doc):
+	"""
+	Generate a name using Prompt option. This simply means the user will have to set the name manually.
+	This is called when the doctype's `autoname` field starts with 'prompt'.
+	"""
+	# set from __newname in save.py
+	if not doc.name:
+		frappe.throw(_("Name not set via prompt"))
+
+
+def _naming_series_autoname(autoname, doc):
+	"""
+	Generate a name using naming series.
+	"""
+	if not doc.naming_series:
+		frappe.throw(frappe._("Naming Series mandatory"))
+
+	name = make_autoname(doc.naming_series+'.#####', '', doc)
+	return name
+
+
+def _concatenate_autoname(autoname, doc):
+	"""
+	Generate a name by concatenating as many fields as required. This is 
+	called when the doctype's `autoname` field starts with 'concatenate:'. The
+	field names are separated by a comma. It is also aware of autoname '.####' 
+	format.
+	"""
+
+	fieldname = autoname[12:]
+	fieldnames = []
+
+	for part in fieldname.split(','):
+		if '.#' in part:
+			fieldnames.append(make_autoname(part, doc))
+		else:
+			fieldnames.append(_field_autoname(part, doc, skip_slicing=True))
+	name = '-'.join(fieldnames)
+	return name
+
+
+def _get_name_from_naming_options(autoname, doc):
```

Comment:
@tundebabzy in frappe we usually the follow the sequence of the most significant function on top and inner functions below so that when you are reviewing a file from top to bottom, its easy to understand the flow. I will update the contributing guidelines for this :)

Can you change the sequence here.

Reply:
Updated: https://github.com/frappe/erpnext/wiki/Contribution-Guidelines